### PR TITLE
[ETL-611] Raw sync lambda

### PR DESCRIPTION
**Pull request by philerooski**
# Raw Sync Lambda

This Lambda verifies that the input and raw S3 buckets are synchronized. It's triggered by a Cloudwatch Events rule at midnight UTC each day.

This is accomplished by verifying that all non-zero sized JSON in each export in the input S3 bucket have a corresponding object in the raw S3 bucket. Because we only download the central directory, located near the end of a zip archive, verification can be done extremely quickly and without needing to download most of the export. If a JSON file from an export is found to not have a corresponding object in the raw bucket, the export is submitted to the raw Lambda (via the dispatch SNS topic) for processing.

### CODE REVIEW

File: src/lambda_function/raw_sync/app.py [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+        key (str): The S3 object key to process.
+        key_format (str): The format of the key, either "raw" or "input".
+        result (dict): The dictionary where keys are appended. For the "raw" format, it is a
+                       nested dictionary structured as result[data_type][cohort]. For "input",
+                       it is structured as result[cohort].
+
+    Returns:
+        None
+    """
+    if not key.endswith("/"):  # Ignore keys that represent "folders"
```
Comment (thomasyu888) [inline src/lambda_function/raw_sync/app.py:73]: Nit: Is there a better way to achieve this?

File: src/lambda_function/raw_sync/app.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+                    for part in key_components
+                    if part.startswith("cohort=")
+                )
+                result[data_type][cohort].append(key)
+            except StopIteration:
+                # Skip keys that don't match the expected pattern
+                return
+        elif key_format == "input" and len(key_components) == 3:
+            cohort = key_components[1]
+            result[cohort].append(key)
```
Comment (thomasyu888) [inline src/lambda_function/raw_sync/app.py:93]: Does this function not return anything?  Are you taking advantage of the mutable dict that you pass in through memory?  I feel like that adds vulnerability in the code

File: src/lambda_function/raw_sync/app.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+    paginator = s3_client.get_paginator("list_objects_v2")
+    response_iterator = paginator.paginate(Bucket=bucket, Prefix=key_prefix)
+    if key_format == "raw":
+        result = defaultdict(lambda: defaultdict(list))
+    elif key_format == "input":
+        result = defaultdict(list)
+    for response in response_iterator:
+        for obj in response.get("Contents", []):
+            key = obj["Key"]
+            append_s3_key(
```
Comment (thomasyu888) [inline src/lambda_function/raw_sync/app.py:163]: It may be better to return the values to append here

**Review by thomasyu888 — APPROVED** [review]
🔥 LGTM, I'm going to let @rxu17 do a review here, but I took a look at the tests and they look good.

The most concern I have is around data quality and making sure that this change doesn't introduce bugs in the production data.  I wonder if we need GX truly running in production so we can be certain changes don't add unintended bugs.

File: src/lambda_function/raw_sync/app.py [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+
+    Args:
+        body (bytes): The byte content from which to extract EOCD fields.
+        eocd_offset (int): The offset position of the EOCD signature in the body.
+
+    Returns:
+        tuple: A tuple containing (central_directory_offset, central_directory_size).
+            Both are int type.
+    """
+    eocd_fields = struct.unpack("<4s4H2LH", body[eocd_offset : eocd_offset + 22])
```
Comment (rxu17) [inline src/lambda_function/raw_sync/app.py:263]: Is it possible to get an invalid `eocd_offset` (not -1 but something else)? Do we have handling for that?

File: src/lambda_function/raw_sync/app.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+        if adjusted_range_size > total_size * 2:
+            logger.error(
+                "Did not find an end of central directory record in "
+                f"s3://{bucket}/{key}"
+            )
+            return []
+        logger.warning(
+            f"Calling this function recursively with `range_size` = {adjusted_range_size}"
+        )
+        return list_files_in_archive(
```
Comment (rxu17) [inline src/lambda_function/raw_sync/app.py:380]: Given the current recursion limit in python being 1000, in the scenario that we hit it, are we concerned about `range_size` being multiplied 1000 times/large?

File: tests/test_lambda_raw_sync.py
Comment (rxu17) [inline tests/test_lambda_raw_sync.py:1]: A few tests that I think would be good to have (I might have missed some tests so please let me know):

- A test that checks to see if StopIteration exception is met in `append_s3_key` function and the dict of s3 keys is not changed
- Two more tests for `match_corresponding_raw_object` function that tests when you'd get a None return so there's comprehensive coverage:
  - When cohort not found
  - When data_type not found
- A test for `list_files_in_archive` function when we end up with empty list of files

File: src/lambda_function/raw_sync/app.py [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+    raw_keys = list_s3_objects(
+        s3_client=s3_client,
+        bucket=raw_bucket,
+        key_prefix=raw_key_prefix,
+        key_format="raw",
+    )
+    for export_key in sum(export_keys.values(), []):
+        # input bucket keys are formatted like `{namespace}/{cohort}/{export_basename}`
+        namespace, cohort = export_key.split("/")[:2]
+        file_list = list_files_in_archive(
```
Comment (rxu17) [inline src/lambda_function/raw_sync/app.py:483]: technically, should this ever be []? Should we be throwing some warning?

**Review by rxu17 — COMMENTED** [review]
Just did a first pass and had some comments/questions. Great work so far!

File: src/lambda_function/raw_sync/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# Raw Sync Lambda
+
+The raw sync Lambda ensures that the input and raw S3 buckets are synchronized
+by verifying that all non-zero sized JSON in each export in the input bucket
+have a corresponding object in the raw bucket.
+
+If a JSON file from an export is found to not have a corresponding object in the raw bucket,
+the file is submitted to the raw Lambda (via the dispatch SNS topic) for processing.
```
[the remark this answers is not in the document]
Comment (philerooski) [reply src/lambda_function/raw_sync/README.md:8]: That's a possibility. Assuming the raw Lambda is working correctly, that could mean a few things:

1. The raw Lambda was throttled, or it failed for normal resource constraint reasons, in which case we do want to re-submit. 
2. The file is unable to be processed by the raw Lambda in its current configuration. Probably this means we've encountered an extremely large JSON file, probably ~120-150 GB uncompressed, or 3 times the size of anything we've received before. If we don't handle this case, the consequences are that we will submit this behemoth for processing once a day where it will timeout after 15 minutes. Alternatively, perhaps a non-zip file or some other corrupted file is accidentally deposited in the input bucket, which would be ignored by the raw-sync Lambda after it determines it's not a zip-file (although this process would also repeat each day until the corrupted file is removed). We don't want to re-submit the file, but I can't think of a graceful way for the raw-sync lambda to differentiate between this case and case (1).

In short... I can't think of a way for the raw-sync Lambda to gracefully handle files which simply can't be processed by raw Lambda. It could help to have raw-sync Lambda publish to an SNS topic which sends us an email when it encounters a file which is out-of-sync, at which point we could manually intervene to see what happened. In the interim we could watch the raw Lambda's DLQ to see if there are any files which repeatedly fail to be processed.

File: src/lambda_function/raw_sync/app.py (hunk 1 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:73]: Not that I know of.

File: src/lambda_function/raw_sync/app.py (hunk 2 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:93]: The function modifies the `result` dict in-memory.

> I feel like that adds vulnerability in the code

How so? I could make a copy of the dict, update the copy, and then return -- but what's the benefit?

File: src/lambda_function/raw_sync/app.py (hunk 4 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:263]: Not according to [the docs](https://docs.python.org/3/library/stdtypes.html#bytes.rfind). See `determine_eocd_offset` for where that method is used.

File: src/lambda_function/raw_sync/app.py (hunk 5 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:380]: That would probably mean a very large file (which is almost certainly not a zip file) was dropped in the input bucket. How large? 32*2^1000 KB, or 3.5e290 petabytes, much larger than any object in S3 could ever be.

So we will never reach the recursion limit and would instead timeout after 15 minutes of trying to download enough of the object to find the EOCD marker.

File: src/lambda_function/raw_sync/app.py (hunk 6 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:483]: CE shouldn't be sending us empty zip files, but agnostic of what's going on in the broader context of the pipeline, a function like `list_files_in_archive` should return an empty list if there are no files in the archive.

If we want a mechanism to alert us of suspiciously empty exports or other data validation issues with data in the input bucket, I don't think the raw-sync Lambda is the right context for that.

File: tests/test_lambda_raw_sync.py [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+            "namespace/cohort_one/object_two",
+        ],
+        "cohort_two": ["namespace/cohort_two/object_one"],
+    }
+
+    result = app.list_s3_objects(s3_client, bucket_name, key_prefix, key_format)
+    assert result == expected_output, f"Expected {expected_output}, but got {result}"
+
+
+def test_match_corresponding_raw_object(mocked_raw_keys):
```
[the remark this answers is not in the document]
Comment (philerooski) [reply tests/test_lambda_raw_sync.py:423]: Sure, although I'm not sure what you mean by run the tests simultaneously.

[the remark this answers is not in the document]
Comment (rxu17) [reply tests/test_lambda_raw_sync.py:423]: Oh I just meant when pytest is called, the tests will be run independent of each other. Here it's dependent on the previous test passing before the following condition can be run.

File: src/lambda_function/raw_sync/app.py (hunk 5 above)
Comment (rxu17) [reply src/lambda_function/raw_sync/app.py:380]: Could there be a scenario where `eocd_offset` was determined incorrectly (-1) for a non-large file?

File: src/lambda_function/raw_sync/app.py (hunk 6 above)
Comment (rxu17) [reply src/lambda_function/raw_sync/app.py:483]: I guess I'm also thinking about if somehow the logic in `list_files_in_archive` was inaccurate after updating it or what not, leading to a [] file. Should we flag that in `list_files_in_archive`?

File: src/lambda_function/raw_sync/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rxu17) [reply src/lambda_function/raw_sync/README.md:8]: I can see how the 2nd scenario would be difficult to catch. I don't remember the retry limit but yeah if it ends up in the DLQ, that can be a good area to check in the interim.

File: src/lambda_function/raw_sync/app.py (hunk 5 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:380]: It can't be a zip file without the EOCD byte marker, so it's not possible to have a false negative. Is that what you mean?

File: src/lambda_function/raw_sync/app.py (hunk 6 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:483]: Are you thinking of what if there is a false negative? We return an empty list in two cases:

1. The EOCD marker is not found and the next `range_size` value we want to attempt is greater than 2x the object size, meaning we just downloaded the entire object and didn't find an EOCD marker. Hence, this is not a valid zip file and there's no file list to return.
2. We find an EOCD marker (this is a valid zip file), but either the zip file is empty, or it contains files which we don't want to process anyway according to our conditions:
```
"/" not in zip_info.filename  # necessary for pilot data only
and "Manifest" not in zip_info.filename
and zip_info.file_size > 0
```

We do `logger.error` in the case of (1), because we shouldn't pass non-zip files to `list_files_in_archive`. It could be useful to do something like `logger.warning` if `file_list` is still empty after (2), because an empty zip-file or a zip-file containing no useful data is pretty unusual, although it doesn't prevent `list_files_in_archive` from doing exactly what we expect of it. If there is a bug in the logic that the tests didn't catch, than a `logger.warning` could help us detect that.

File: src/lambda_function/raw_sync/app.py (hunk 2 above)
Comment (philerooski) [reply src/lambda_function/raw_sync/app.py:93]: Actually, I did think of a small benefit. Returning a copy makes the variable update in `list_s3_objects` more explicit, and creating a shallow copy within the function can be done without any additional imports and with just one extra line of code.

File: src/lambda_function/raw_sync/app.py (hunk 5 above)
Comment (rxu17) [reply src/lambda_function/raw_sync/app.py:380]: Yup

File: src/lambda_function/raw_sync/app.py (hunk 6 above)
Comment (rxu17) [reply src/lambda_function/raw_sync/app.py:491]: Sounds good to me!

**Review by rxu17 — APPROVED** [review]
LGTM!

File: src/lambda_function/raw_sync/app.py (hunk 2 above)
Comment (thomasyu888) [reply src/lambda_function/raw_sync/app.py:94]: This is probably more relevant when making mutable default arguments, but I prefer the way you've written it to make the function self contained
https://docs.python-guide.org/writing/gotchas/#:~:text=Python's%20default%20arguments%20are%20evaluated,to%20the%20function%20as%20well.
